amazon/ion/ioncmodule.c

```diff
@@ -0,0 +1,1369 @@
+    Py_DECREF(sequence_as_stream);
+    Py_DECREF(tuple_as_sexp);
+    PyObject* exception = NULL;
+    if (err == IERR_INVALID_STATE) {
+        exception = PyErr_Format(PyExc_TypeError, "%s", _err_msg);
+    }
+    else {
+        exception = PyErr_Format(_ion_exception_cls, "%s %s", ion_error_to_str(err), _err_msg);
+    }
+
+    _err_msg[0] = '\0';
+    return exception;
+}
+
+
+/******************************************************************************
+*       Read/Load APIs                                                        *
+******************************************************************************/
+
+
+static PyObject* ionc_get_timestamp_precision(int precision) {
+    int precision_index = -1;
+    while (precision) {
+        precision_index++;
+        precision = precision >> 1;
+    }
+    return py_ion_timestamp_precision_table[precision_index];
+}
+
+static iERR ionc_read_timestamp(hREADER hreader, PyObject** timestamp_out) {
+    iENTER;
+    ION_TIMESTAMP timestamp_value;
+    PyObject* timestamp_args = NULL;
+    IONCHECK(ion_reader_read_timestamp(hreader, &timestamp_value));
+    int precision;
+    IONCHECK(ion_timestamp_get_precision(&timestamp_value, &precision));
+    if (precision < ION_TS_YEAR) {
+        _FAILWITHMSG(IERR_INVALID_TIMESTAMP, "Found a timestamp with less than year precision.");
+    }
+    timestamp_args = PyDict_New();
+    PyObject* py_precision = ionc_get_timestamp_precision(precision);
+    PyDict_SetItemString(timestamp_args, "precision", py_precision);
+    BOOL has_local_offset;
+    IONCHECK(ion_timestamp_has_local_offset(&timestamp_value, &has_local_offset));
+
+    if (has_local_offset) {
+        int off_minutes, off_hours;
+        IONCHECK(ion_timestamp_get_local_offset(&timestamp_value, &off_minutes));
+        off_hours = off_minutes / 60;
+        off_minutes = off_minutes % 60;
+        // Bounds checking is performed in python.
+        PyDict_SetItemString(timestamp_args, "off_hours", PyInt_FromLong(off_hours));
+        PyDict_SetItemString(timestamp_args, "off_minutes", PyInt_FromLong(off_minutes));
+    }
+
+    switch (precision) {
+        case ION_TS_FRAC:
+        {
+            decQuad fraction = timestamp_value.fraction;
+            int32_t fractional_precision = decQuadGetExponent(&fraction);
+            if (fractional_precision > 0) {
+                _FAILWITHMSG(IERR_INVALID_TIMESTAMP, "Timestamp fractional precision cannot be a positive number.");
+            }
+            fractional_precision = fractional_precision * -1;
+
+            decQuad tmp;
+            decQuadScaleB(&fraction, &fraction, decQuadFromInt32(&tmp, MICROSECOND_DIGITS), &dec_context);
+            int32_t microsecond = decQuadToInt32Exact(&fraction, &dec_context, DEC_ROUND_DOWN);
+            if (fractional_precision > MICROSECOND_DIGITS) {
+                // Python only supports up to microsecond precision
+                fractional_precision = MICROSECOND_DIGITS;
```

Comment:
I didn't change too much things for time stamp API, this is one of the few changes I made.

[Before](https://github.com/cheqianh/ion-python/blob/43bf9243c9b145f8e2baca32624f693bfb3d87ab/amazon/ion/ioncmodule.c#L733) we throw an error if the timestamp value has more than `MICROSECOND_DIGITS` precision. In order to match c spec and not fail tests, now we keep `MICROSECOND_DIGITS` digits and set `DEC_ROUND_DOWN`.

e.g. by `DEC_ROUND_DOWN`, 0.1234567 s will be round to 0.123456